tests/unit/bootstrap_record_test.cc

```diff
@@ -0,0 +1,445 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+#include "fs/bootstrap_record.hh"
+#include "fs/cluster.hh"
+
+#include <boost/crc.hpp>
+#include <cstring>
+#include <seastar/core/print.hh>
+#include <seastar/fs/block_device.hh>
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/test_runner.hh>
+#include <seastar/testing/thread_test_case.hh>
+
+using namespace seastar;
+using namespace seastar::fs;
+
+namespace {
+
+class mock_block_device_impl : public block_device_impl {
+public:
+    using buf_type = basic_sstring<uint8_t, size_t, 32, false>;
+    buf_type buf;
+    ~mock_block_device_impl() override = default;
+
+    future<size_t> write(uint64_t pos, const void* buffer, size_t len, const io_priority_class&) override {
+        if (buf.size() < pos + len)
+            buf.resize(pos + len);
+        std::memcpy(buf.data() + pos, buffer, len);
+        return make_ready_future<size_t>(len);
+    }
+
+    future<size_t> read(uint64_t pos, void* buffer, size_t len, const io_priority_class&) noexcept override {
+        std::memcpy(buffer, buf.c_str() + pos, len);
+        return make_ready_future<size_t>(len);
+    }
+
+    future<> flush() noexcept override {
+        return make_ready_future<>();
+    }
+
+    future<> close() noexcept override {
+        return make_ready_future<>();
+    }
+};
+
+inline std::vector<bootstrap_record::shard_info> prepare_valid_shards_info(uint32_t size) {
+    std::vector<bootstrap_record::shard_info> ret(size);
+    cluster_id_t curr = 1;
+    for (bootstrap_record::shard_info& info : ret) {
+        info.available_clusters = {curr, curr + 1};
+        info.metadata_cluster = curr;
+        curr++;
+    }
+    return ret;
+};
+
+inline void repair_crc32(shared_ptr<mock_block_device_impl> dev_impl) noexcept {
+    auto crc32 = [](const uint8_t* buff, size_t len) {
```

Comment:
```suggestion
```

Don't repeat yourself and move
```cpp 
inline uint32_t crc32(const void* buff, size_t len) noexcept;
```
from `src/fs/bootstrap_record.cc` to `src/fs/bootstrap_record.hh`

Reply:
I am not a fan of that solution because crc32 is quite unrelated to bootstrap record. I would rather move that function into another header file like crc.hh. What do you think about that solution?

Reply:
Sure, sounds good.